File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.java
import java.util.*;

class Solution {

    static List<Integer>[] tree;
    static int n, m, answer;


    public int solution(int[] info, int[][] edges) {
        n = info.length;
        m = edges.length;

        boolean[] vis = new boolean[n];
       tree = new ArrayList[n];
        for(int i = 0; i < n; i++)
            tree[i] = new ArrayList<>();

        for(int i = 0; i < m; i++) 
            tree[edges[i][0]].add(edges[i][1]);

        dfs(0, 0, 0, 0, info);

        return answer;
    }

    static void dfs(int node, int sheep, int wolf, int vis, int[] info) {

        if(info[node] == 0) {
            sheep++;
        }else {
            wolf++;
        }

        for(int nxt : tree[node])
            vis |= (1<<nxt);

        if(sheep == wolf)
            return;

        answer = answer > sheep ? answer : sheep;

        for(int i = 0; i < n; i++) {
            if((vis & (1<<i)) != 0) {
                vis ^= (1 << i);
                dfs(i, sheep, wolf, vis, info);
                vis ^= (1<< i);
            }
        }

    }
}

File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.js
function solution(info, edges) {
  var answer = 0;


  const nodeCnt = info.length;
  const graph = Array.from({ length: nodeCnt }, () => []);

  for (let [a, b] of edges) {
    graph[a].push(b);
  }

  let maxSheepCnt = -1;

  const dfs = (x, sheepCnt, wolfCnt, visited) => {
    if (sheepCnt <= wolfCnt) return;
    maxSheepCnt = maxSheepCnt > sheepCnt ? maxSheepCnt : sheepCnt;

    for (let i = 0; i < graph[x].length; i++) {
      const nx = graph[x][i];
      visited = visited | (1 << nx);
    }

    for (let i = 0; i < nodeCnt; i++) {
      if ((visited & (1 << i)) === 0) continue;

      visited = visited ^ (1 << i);

      if (info[i] === 1) {
        dfs(i, sheepCnt, wolfCnt + 1, visited);
      } else if (info[i] === 0) {
        dfs(i, sheepCnt + 1, wolfCnt, visited);
      }
      visited = visited ^ (1 << i);
    }
  };

  dfs(0, 1, 0, 0);

  answer = maxSheepCnt;
  return answer;
}